### PR TITLE
feature/file_upload_api

### PR DESCRIPTION
**Pull request by jrouzierinverse**
# Description
Allow users to upload files from the filesystem.

# Impacts
The admin and the pfperl-api.
# Delete branch after merge
YES

# Checklist
(REQUIRED) - [yes, no or n/a]
- [ ] Document the feature
- [x] Add unit tests
- [ ] Add acceptance tests (TestLink)

# NEWS file entries
## Enhancements
* Allow users to upload files from the admin.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -34,6 +34,7 @@ conf/pfdhcp.conf
 conf/etcd.conf.yml
 conf/ssl.conf
 conf/template_switches.conf
+conf/uploads/*
```
Comment (julsemaan) [inline .gitignore:37]: This directory needs to be packaged in Debian and RHEL

File: html/pfappserver/lib/pfappserver/Form/Config/PKI_Provider/packetfence_pki.pm [hunk 2]
```diff
@@ -144,13 +144,23 @@ has_field 'organization' => (
...
 );
 
+has_field 'ca_cert_path_upload' => (
+   type => 'PathUpload',
+   accessor => 'ca_cert_path',
+   config_prefix => '.crt',
+   label => 'File Upload',
+   required => 0,
+   upload_namespace => 'pki',
+   element_class => ['input-xxlarge'],
```
Comment (julsemaan) [inline html/pfappserver/lib/pfappserver/Form/Config/PKI_Provider/packetfence_pki.pm]: All these element_class and label don't seem relevant to me since they are not consumed by the frontend nor included in the OPTIONS

Comment (jrouzierinverse) [reply html/pfappserver/lib/pfappserver/Form/Config/PKI_Provider/packetfence_pki.pm]: An artifact from copying and pasting code.
When should I take the time to remove it for all the forms?

Comment (julsemaan) [reply html/pfappserver/lib/pfappserver/Form/Config/PKI_Provider/packetfence_pki.pm]: I don't think we should remove them from all the forms now but at least not add them to the new fields we're adding into the forms

**Comment by julsemaan** [comment]
@jrouzierinverse, it has failed to build:
https://gitlab.com/inverse-inc/packetfence/-/jobs/3218161614

**Comment by julsemaan** [comment]
Very nice work on this!

I have 2 issues I have to report which I believe @jrouzierinverse will need to take care of

1. The permissions on the uploaded files are incorrect. The directories and files are not chowned to pf after the upload
2. Creating a SAML source will uploaded files fails (updating the file in the file1 source worked so unsure where the specific issue is):
image

**Comment by julsemaan** [comment]
I'm now able to create the SAML source but still 2 issues:
- backend for @jrouzierinverse  -> The owner+group on the files uploaded in the SAML source are still wrong

- frontend for @satkunas  -> The fields containing the paths after uploading in the SAML source are all empty but they're actually filled in the config. This happens for SAML but not for other sections like provisioners. Reloading the page has the data appear in the fields

image

**Comment by jrouzierinverse** [comment]
This has been fixed.

**Comment by satkunas** [comment]
Additional fixed pushed for UI.

**Comment by satkunas** [comment]
Screen Shot 2022-10-26 at 4 51 02 PM

@julsemaan do we want to make the last 2 fields (android provisioner) mandatory?

**Comment by julsemaan** [comment]
> Screen Shot 2022-10-26 at 4 51 02 PM
> 
> @julsemaan do we want to make the last 2 fields (android provisioner) mandatory?

They are mandatory only when WPA2/PEAP is selected and maybe also when WPA2/TLS is selected. We must ensure that they are not mandatory in the other cases

**Comment by satkunas** [comment]
> > Screen Shot 2022-10-26 at 4 51 02 PM
> > @julsemaan do we want to make the last 2 fields (android provisioner) mandatory?
> 
> They are mandatory only when WPA2/PEAP is selected and maybe also when WPA2/TLS is selected. We must ensure that they are not mandatory in the other cases

Tip: Validation errors only block the submission of the form if they're visible in the DOM.

**Comment by julsemaan** [comment]
@jrouzierinverse, all we're missing is the cleanup of the useless tags and labels in the forms.

All is working, we're good to merge right after that

**Comment by jrouzierinverse** [comment]
Done

**Comment by julsemaan** [comment]
> Done

I'm guessing my comment was unclear. So it's the removal of all unused, including the labels which are in the frontend now

**Comment by jrouzierinverse** [comment]
Done

**Comment by julsemaan** [comment]
Merging 💪 Great work!
